### PR TITLE
helm: Added Pod Security Policy

### LINKED ISSUE #404: Add pod security policy to the helm chart
**Describe the request**
Right now we have a pod security policy enabled in our k8s clusters that prevent the creation of certain volumes and allow only certain capabilities, etc.. this is a blocker to deploy secrets-store-csi-driver-provider-azure properly so we had to create the following pod security policy manually.

```
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: secrets-store
spec:
  seLinux:
    rule: RunAsAny
  privileged: true
  volumes:
    - hostPath
    - secret
  hostNetwork: true
  hostPorts:
  - min: 0
    max: 65535
  fsGroup:
    rule: RunAsAny
  runAsUser:
    rule: RunAsAny
  supplementalGroups:
    rule: RunAsAny
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: psp-secrets-store-role
  namespace: NAME
rules:
- apiGroups: ['extensions']
  resources: ['podsecuritypolicies']
  verbs:     ['use']
  resourceNames:
  - secrets-store
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: psp-secrets-store-role-binding
  namespace: NAME
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: psp-secrets-store-role
subjects:
- kind: ServiceAccount
  name: csi-secrets-store-provider-azure
  namespace: NAME
```

**Explain why AAD Pod Identity needs it**
To deploy secrets-store-csi-driver-provider-azure successfully in k8s cluster with a restricted pod-security policy enabled.

**Describe the solution you'd like**
Merge the pod security policy mentioned above in the software helm chart, and keep it updated with the new software releases.

**Describe alternatives you've considered**
the other alternative is to create the pod security policy manually, which is not good if more capabilities are needed with new sw releases and we have always to update the yaml manually

**Environment:**

- Secrets Store CSI Driver version: (use the image tag): latest
- Kubernetes version: (use `kubectl version`): v1.18.8

### PR DESCRIPTION
**Pull request by pierluigilenoci**
**Reason for Change**:
We want to deploy secrets-store-csi-driver-provider-azure successfully in k8s cluster with a restricted pod-security policy enabled.

**Requirements**

- [x] squashed commits
- [x] included documentation
- [x] added unit tests and e2e tests (if applicable).

**Issue Fixed**:
Fixes #404 

**Does this change contain code from or inspired by another project?**

- [ ] Yes
- [x] No

**Special Notes for Reviewers**:

This is inspired by @mo-saeed code from #404

### CODE REVIEW

**Review by mo-saeed — APPROVED** [review]

File: manifest_staging/charts/csi-secrets-store-provider-azure/README.md [hunk 1]
```diff
@@ -103,5 +103,6 @@ The following table lists the configurable parameters of the csi-secrets-store-p
...
 | `secrets-store-csi-driver.rotationPollInterval`                  | Secret rotation poll interval duration                                                                                        | `2m`                                                                                             |
 | `rbac.install`                                                   | Install default service account                                                                                               | true                                                                                             |
+| `rbac.pspEnabled`                                                | If `true`, create and use a restricted pod security policy for Secrets Store CSI Driver pod(s)                                                                                       | true                                                                                             |
```
Comment (aramase) [inline manifest_staging/charts/csi-secrets-store-provider-azure/README.md:106]: Default should be false as that's the current behavior

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.rbac.pspEnabled }}
+apiVersion: policy/v1beta1
+kind: PodSecurityPolicy
+metadata:
+  name: {{ template "sscdpa.fullname" . }}
```
Comment (aramase) [inline manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:5]: If the release name is used for psp and the release name begins with `z-*` for instance, then the pods will get blocked by the restricted policy if it exists. Should the policy instead have a custom name?

cc @sozercan

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+kind: Role
+metadata:
+  name: {{ template "sscdpa.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+rules:
+  - apiGroups: [ 'extensions' ]
+    resources: [ 'podsecuritypolicies' ]
+    verbs: [ 'use' ]
+    resourceNames:
+      - secrets-store
```
Comment (sozercan) [inline manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml:12]: i think this needs to be `{{ template "sscdpa.psp.fullname" . }}`

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml [hunk 4]
```diff
@@ -0,0 +1,13 @@
+{{- if .Values.rbac.pspEnabled }}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: Role
+metadata:
+  name: {{ template "sscdpa.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+rules:
+  - apiGroups: [ 'extensions' ]
```
Comment (sozercan) [inline manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml:8]:
```suggestion
  - apiGroups: [ 'policy' ]
```

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml (hunk 3 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml:12]: fixed

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml (hunk 4 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml]: fixed

**Comment by pierluigilenoci** [comment]
@aramase @sozercan can we push this forward? What is needed to do that?

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{{- if .Values.rbac.pspEnabled }}
```
Comment (aramase) [inline manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]:
```yaml
{{- if .Values.rbac.pspEnabled }}
apiVersion: policy/v1beta1
kind: PodSecurityPolicy
metadata:
  name: {{ template "sscdpa.psp.fullname" . }}
spec:
  hostNetwork: true
  seLinux:
    rule: RunAsAny
  privileged: {{ .Values.linux.privileged }}
  volumes:
    - hostPath
    - secret
  hostPorts:
    - min: 0
      max: 65535
  fsGroup:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
  runAsUser:
    rule: RunAsAny
  supplementalGroups:
    ranges:
    - max: 65535
      min: 1
    rule: MustRunAs
{{- end }}
```

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml (hunk 3 above)
Comment (aramase) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml:12]: This is still not fixed

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/rolebinding.yaml [hunk 6]
```diff
@@ -0,0 +1,15 @@
+{{- if .Values.rbac.pspEnabled }}
+apiVersion: rbac.authorization.k8s.io/v1
+kind: RoleBinding
+metadata:
+  name: {{ template "sscdpa.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: Role
+  name: {{ template "sscdpa.fullname" . }}
```
Comment (aramase) [inline manifest_staging/charts/csi-secrets-store-provider-azure/templates/rolebinding.yaml:10]: This needs to be `name: {{ template "sscdpa.psp.fullname" . }}`

**Review by aramase — CHANGES_REQUESTED** [review]
@pierluigilenoci Added comments. PTAL and also rebase the PR when you get a chance.

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml (hunk 5 above)
Comment (aramase) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]: Also, please include the default labels in the `PodSecurityPolicy`, `Role` and `RoleBinding`

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml (hunk 3 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/role.yaml]: Done

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/rolebinding.yaml (hunk 6 above)
Comment (pierluigilenoci) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/rolebinding.yaml]: Done

**Comment by pierluigilenoci** [comment]
@aramase sorry it took me a while but last week I just couldn't. 👍🏻

**Review by aramase — APPROVED** [review]

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml (hunk 5 above)
Comment (sozercan) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]: @aramase should this match deployment PSP here? https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/deployment/pod-security-policy.yaml#L57-L73

Comment (aramase) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]: @sozercan Apart from the `allowedHostPaths `, the rest of the spec is very similar here. Imposing the different host paths can get tedious if users want to mount custom env file or more. WDYT? 

Also, that file seems to be a little outdated. We should update it in a follow up PR.

**Review by sozercan — APPROVED** [review]

File: manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml (hunk 5 above)
Comment (sozercan) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]: sounds good. we can have a follow up to address that.

Comment (aramase) [reply manifest_staging/charts/csi-secrets-store-provider-azure/templates/podsecuritypolicy.yaml:1]: Opened https://github.com/Azure/secrets-store-csi-driver-provider-azure/issues/479 to track the change.

**Pull request merged**